### PR TITLE
Llu/welford complex

### PR DESCRIPTION
**Pull request by liqiangxl**
solves https://github.com/csarofeen/pytorch/issues/2527.
Call Welford twice, one for the real part and the other for the imaginary part. Then calculate the mean (complex number), var (real number), and n.
```
      WelfordResult real_part =
          Welford(real(squeezed), reduction_axes, init_avg, init_var, init_N);
      WelfordResult imag_part =
          Welford(imag(squeezed), reduction_axes, init_avg, init_var, init_N);
      TensorView* out_avg = complex(real_part.avg, imag_part.avg);
      TensorView* out_var = add(real_part.var_sum, imag_part.var_sum);
      TensorView* out_N = real_part.n;
      return WelfordResult(out_avg, out_var, out_N, false);
```

### CODE REVIEW

File: csrc/type.h [hunk 1]
```diff
@@ -341,7 +341,11 @@ enum class BinaryOpType {
   // is boolean op. These ops also don't work on floating point inputs.
   And,
   Or,
-  Xor
+  Xor,
+
+  // generate complex from real and imaginary parts
+  ComplexFloat,
+  ComplexDouble
```
Comment (zasdfgbnm) [inline csrc/type.h:348]: Do we need two binary op type? PyTorch only has a single complex https://pytorch.org/docs/stable/generated/torch.complex.html?highlight=complex#torch.complex

File: csrc/type.cpp [hunk 2]
```diff
@@ -379,6 +379,10 @@ static const char* binary_op_type2string(BinaryOpType t) {
       return "remainder";
     case BinaryOpType::Sub:
       return "sub";
+    case BinaryOpType::ComplexFloat:
+      return "std::complex<float>";
+    case BinaryOpType::ComplexDouble:
+      return "std::complex<double>";
```
Comment (zasdfgbnm) [inline csrc/type.cpp:385]: I think we should only have a single binary op type, and we can use [the CTAD of C++17](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction) to infer the returning type

File: csrc/ops/arith.cpp [hunk 3]
```diff
@@ -1744,7 +1767,21 @@ WelfordResult Welford(
   }
 
   if (!reduction_axes.empty()) {
-    return WelfordRaw(squeezed, reduction_axes, init_avg, init_var, init_N);
+    DataType dtype = tv->getDataType().value();
+    if (dtype == DataType::ComplexFloat || dtype == DataType::ComplexDouble) {
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:1771]: nit: use this: https://github.com/NVIDIA/Fuser/blob/9f9ba02d2d4aab6587ca76c3869baec4299c9459/csrc/type.h#L200

File: csrc/ops/arith.cpp [hunk 4]
```diff
@@ -657,6 +657,31 @@ TensorView* imag(TensorView* tv) {
...
+  BinaryOpType complex_op;
+  if (dtype == DataType::Double) {
+    complex_dtype = DataType::ComplexDouble;
+    complex_op = BinaryOpType::ComplexDouble;
+  } else if (dtype == DataType::Float) {
+    complex_dtype = DataType::ComplexFloat;
+    complex_op = BinaryOpType::ComplexFloat;
+  } else {
+    TORCH_CHECK(false, "complex() only supports float and double types");
+  }
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:676]: nit: please add a `getComplexTypeFromType` similar to https://github.com/NVIDIA/Fuser/blob/9f9ba02d2d4aab6587ca76c3869baec4299c9459/csrc/type.h#L208

File: csrc/ops/arith.cpp (hunk 3 above)
Comment (liqiangxl) [reply csrc/ops/arith.cpp:1771]: done.

File: csrc/ops/arith.cpp (hunk 4 above)
Comment (liqiangxl) [reply csrc/ops/arith.cpp:676]: done.

File: csrc/type.h (hunk 1 above)
Comment (liqiangxl) [reply csrc/type.h:348]: changed to one.

File: csrc/type.cpp (hunk 2 above)
Comment (liqiangxl) [reply csrc/type.cpp:385]: CTAD works in this case. At the beginning I used std::compex but doesn't work, maybe due to other reason.

**Pull request merged**